# Q4
## Coding workspace for LP solving in question 4 of homework 9

#### S - Original

In [ ]:
# Code for Solving the Original Problem
import pulp

# Create a LP maximization problem
prob = pulp.LpProblem("Maximization_Problem", pulp.LpMaximize)

# Define decision variables
x1 = pulp.LpVariable('x1', lowBound=0, upBound=1, cat='Continuous')
x2 = pulp.LpVariable('x2', lowBound=0, upBound=1, cat='Continuous')
x3 = pulp.LpVariable('x3', lowBound=0, upBound=1, cat='Continuous')
x4 = pulp.LpVariable('x4', lowBound=0, upBound=1, cat='Continuous')

# Objective function
prob += 9*x1 + 5*x2 + 6*x3 + 4*x4, "Objective"

# Constraints
prob += 6*x1 + 3*x2 + 5*x3 + 2*x4 <= 10, "Constraint 1"
prob += x3 + x4 <= 1, "Constraint 2"
prob += x3 - x1 <= 0, "Constraint 3"
prob += x4 - x2 <= 0, "Constraint 4"

# Solve the problem
prob.solve()

# Print the results
print("**** S - Original Problem ****")
print("Status:", pulp.LpStatus[prob.status])
for variable in prob.variables():
    print(f"{variable.name} = {variable.varValue}")
print("Objective value:", pulp.value(prob.objective))

**** S - Original Problem ****
Status: Optimal
x1 = 0.83333333
x2 = 1.0
x3 = 0.0
x4 = 1.0
Objective value: 16.49999997


S - Original:
- Fractional
- proceed to branch on x1
- add `x1 = 0` (S1)
- add `x1 = 1` (S2)

#### S - 1 : Add `x1 = 0`

In [ ]:
# Code for Solving the Problem with Constraint S1
import pulp

# Create a LP maximization problem
prob = pulp.LpProblem("Maximization_Problem", pulp.LpMaximize)

# Define decision variables
x1 = pulp.LpVariable('x1', lowBound=0, upBound=1, cat='Continuous')
x2 = pulp.LpVariable('x2', lowBound=0, upBound=1, cat='Continuous')
x3 = pulp.LpVariable('x3', lowBound=0, upBound=1, cat='Continuous')
x4 = pulp.LpVariable('x4', lowBound=0, upBound=1, cat='Continuous')

# Objective function
prob += 9*x1 + 5*x2 + 6*x3 + 4*x4, "Objective"

# Constraints
prob += 6*x1 + 3*x2 + 5*x3 + 2*x4 <= 10, "Constraint 1"
prob += x3 + x4 <= 1, "Constraint 2"
prob += x3 - x1 <= 0, "Constraint 3"
prob += x4 - x2 <= 0, "Constraint 4"
prob += x1 == 0, "Constraint S1"

# Solve the problem
prob.solve()

# Print the results
print("**** S - 1 : x1 == 0 ****")
print("Status:", pulp.LpStatus[prob.status])
for variable in prob.variables():
    print(f"{variable.name} = {variable.varValue}")
print("Objective value:", pulp.value(prob.objective))

**** S - 1 : x1 == 0 ****
Status: Optimal
x1 = 0.0
x2 = 1.0
x3 = 0.0
x4 = 1.0
Objective value: 9.0


S - 1 `x1 = 0`:
- Integer
- Current Lbound: `9`
- no need for further branching

#### S2 - Add `x1 = 1`

In [ ]:
# Code for Solving the Problem with Constraint S2
import pulp

# Create a LP maximization problem
prob = pulp.LpProblem("Maximization_Problem", pulp.LpMaximize)

# Define decision variables
x1 = pulp.LpVariable('x1', lowBound=0, upBound=1, cat='Continuous')
x2 = pulp.LpVariable('x2', lowBound=0, upBound=1, cat='Continuous')
x3 = pulp.LpVariable('x3', lowBound=0, upBound=1, cat='Continuous')
x4 = pulp.LpVariable('x4', lowBound=0, upBound=1, cat='Continuous')

# Objective function
prob += 9*x1 + 5*x2 + 6*x3 + 4*x4, "Objective"

# Constraints
prob += 6*x1 + 3*x2 + 5*x3 + 2*x4 <= 10, "Constraint 1"
prob += x3 + x4 <= 1, "Constraint 2"
prob += x3 - x1 <= 0, "Constraint 3"
prob += x4 - x2 <= 0, "Constraint 4"
prob += x1 == 1, "Constraint S2"

# Solve the problem
prob.solve()

# Print the results
print("**** S - 2 : x1 == 1 ****")
print("Status:", pulp.LpStatus[prob.status])
for variable in prob.variables():
    print(f"{variable.name} = {variable.varValue}")
print("Objective value:", pulp.value(prob.objective))

**** S - 2 : x1 == 1 ****
Status: Optimal
x1 = 1.0
x2 = 0.8
x3 = 0.0
x4 = 0.8
Objective value: 16.2


S - 2 `x1 = 1`:
- Fractional
- Current Ubound (LP - relaxed): `16.2`
- Branch on x2
- Add `x2 = 0` (S3)
- Add `x2 = 1` (S4)

#### S3 - Add `x2 = 0`

In [ ]:
# Code for Solving the Problem with Constraint S3
import pulp

# Create a LP maximization problem
prob = pulp.LpProblem("Maximization_Problem", pulp.LpMaximize)

# Define decision variables
x1 = pulp.LpVariable('x1', lowBound=0, upBound=1, cat='Continuous')
x2 = pulp.LpVariable('x2', lowBound=0, upBound=1, cat='Continuous')
x3 = pulp.LpVariable('x3', lowBound=0, upBound=1, cat='Continuous')
x4 = pulp.LpVariable('x4', lowBound=0, upBound=1, cat='Continuous')

# Objective function
prob += 9*x1 + 5*x2 + 6*x3 + 4*x4, "Objective"

# Constraints
prob += 6*x1 + 3*x2 + 5*x3 + 2*x4 <= 10, "Constraint 1"
prob += x3 + x4 <= 1, "Constraint 2"
prob += x3 - x1 <= 0, "Constraint 3"
prob += x4 - x2 <= 0, "Constraint 4"
prob += x1 == 1, "Constraint S2" # branch from S2
prob += x2 == 0, "Constraint S3"

# Solve the problem
prob.solve()

# Print the results
print("**** S - 3 : x2 == 0 ****")
print("Status:", pulp.LpStatus[prob.status])
for variable in prob.variables():
    print(f"{variable.name} = {variable.varValue}")
print("Objective value:", pulp.value(prob.objective))

**** S - 3 : x2 == 0 ****
Status: Optimal
x1 = 1.0
x2 = 0.0
x3 = 0.8
x4 = 0.0
Objective value: 13.8


S - 3 `x2 = 0`:
- Fractional
- Branch on x3
- Add `x3 = 0` (S5)
- Add `x3 = 1` (S6)

#### S4 - Add `x2 = 1`

In [ ]:
# Code for Solving the Problem with Constraint S4
import pulp

# Create a LP maximization problem
prob = pulp.LpProblem("Maximization_Problem", pulp.LpMaximize)

# Define decision variables
x1 = pulp.LpVariable('x1', lowBound=0, upBound=1, cat='Continuous')
x2 = pulp.LpVariable('x2', lowBound=0, upBound=1, cat='Continuous')
x3 = pulp.LpVariable('x3', lowBound=0, upBound=1, cat='Continuous')
x4 = pulp.LpVariable('x4', lowBound=0, upBound=1, cat='Continuous')

# Objective function
prob += 9*x1 + 5*x2 + 6*x3 + 4*x4, "Objective"

# Constraints
prob += 6*x1 + 3*x2 + 5*x3 + 2*x4 <= 10, "Constraint 1"
prob += x3 + x4 <= 1, "Constraint 2"
prob += x3 - x1 <= 0, "Constraint 3"
prob += x4 - x2 <= 0, "Constraint 4"
prob += x1 == 1, "Constraint S2" # branch from S2
prob += x2 == 1, "Constraint S4"

# Solve the problem
prob.solve()

# Print the results
print("**** S - 4 : x2 == 1 ****")
print("Status:", pulp.LpStatus[prob.status])
for variable in prob.variables():
    print(f"{variable.name} = {variable.varValue}")
print("Objective value:", pulp.value(prob.objective))

**** S - 4 : x2 == 1 ****
Status: Optimal
x1 = 1.0
x2 = 1.0
x3 = 0.0
x4 = 0.5
Objective value: 16.0


S - 4 `x2 = 1`:
- Fractional
- Branch on x4
- Add `x4 = 0` (S7)
- Add `x4 = 1` (S8)

#### S5 - Add `x3 = 0`

In [ ]:
# Code for Solving the Problem with Constraint S5
import pulp

# Create a LP maximization problem
prob = pulp.LpProblem("Maximization_Problem", pulp.LpMaximize)

# Define decision variables
x1 = pulp.LpVariable('x1', lowBound=0, upBound=1, cat='Continuous')
x2 = pulp.LpVariable('x2', lowBound=0, upBound=1, cat='Continuous')
x3 = pulp.LpVariable('x3', lowBound=0, upBound=1, cat='Continuous')
x4 = pulp.LpVariable('x4', lowBound=0, upBound=1, cat='Continuous')

# Objective function
prob += 9*x1 + 5*x2 + 6*x3 + 4*x4, "Objective"

# Constraints
prob += 6*x1 + 3*x2 + 5*x3 + 2*x4 <= 10, "Constraint 1"
prob += x3 + x4 <= 1, "Constraint 2"
prob += x3 - x1 <= 0, "Constraint 3"
prob += x4 - x2 <= 0, "Constraint 4"
prob += x1 == 1, "Constraint S2" # branch from S2
prob += x2 == 0, "Constraint S3" # bracnh from S3
prob += x3 == 0, "Constraint S5"

# Solve the problem
prob.solve()

# Print the results
print("**** S - 5 : x3 == 0 ****")
print("Status:", pulp.LpStatus[prob.status])
for variable in prob.variables():
    print(f"{variable.name} = {variable.varValue}")
print("Objective value:", pulp.value(prob.objective))

**** S - 5 : x3 == 0 ****
Status: Optimal
x1 = 1.0
x2 = 0.0
x3 = 0.0
x4 = 0.0
Objective value: 9.0


S - 5 `x3 = 0`:
- Integer
- No need for further branching

#### S6 - Add `x3 = 1`

In [ ]:
# Code for Solving the Problem with Constraint S6
import pulp

# Create a LP maximization problem
prob = pulp.LpProblem("Maximization_Problem", pulp.LpMaximize)

# Define decision variables
x1 = pulp.LpVariable('x1', lowBound=0, upBound=1, cat='Continuous')
x2 = pulp.LpVariable('x2', lowBound=0, upBound=1, cat='Continuous')
x3 = pulp.LpVariable('x3', lowBound=0, upBound=1, cat='Continuous')
x4 = pulp.LpVariable('x4', lowBound=0, upBound=1, cat='Continuous')

# Objective function
prob += 9*x1 + 5*x2 + 6*x3 + 4*x4, "Objective"

# Constraints
prob += 6*x1 + 3*x2 + 5*x3 + 2*x4 <= 10, "Constraint 1"
prob += x3 + x4 <= 1, "Constraint 2"
prob += x3 - x1 <= 0, "Constraint 3"
prob += x4 - x2 <= 0, "Constraint 4"
prob += x1 == 1, "Constraint S2" # branch from S2
prob += x2 == 0, "Constraint S3" # bracnh from S3
prob += x3 == 1, "Constraint S6"

# Solve the problem
prob.solve()

# Print the results
print("**** S - 6 : x3 == 1 ****")
print("Status:", pulp.LpStatus[prob.status])
for variable in prob.variables():
    print(f"{variable.name} = {variable.varValue}")
print("Objective value:", pulp.value(prob.objective))

**** S - 6 : x3 == 1 ****
Status: Infeasible
x1 = 1.0
x2 = 0.0
x3 = 1.0
x4 = -0.5
Objective value: 13.0


S - 6 `x3 = 1`:
- Infeasible
- Prune this branch

#### S7 - Add `x4 = 0`

In [ ]:
# Code for Solving the Problem with Constraint S7
import pulp

# Create a LP maximization problem
prob = pulp.LpProblem("Maximization_Problem", pulp.LpMaximize)

# Define decision variables
x1 = pulp.LpVariable('x1', lowBound=0, upBound=1, cat='Continuous')
x2 = pulp.LpVariable('x2', lowBound=0, upBound=1, cat='Continuous')
x3 = pulp.LpVariable('x3', lowBound=0, upBound=1, cat='Continuous')
x4 = pulp.LpVariable('x4', lowBound=0, upBound=1, cat='Continuous')

# Objective function
prob += 9*x1 + 5*x2 + 6*x3 + 4*x4, "Objective"

# Constraints
prob += 6*x1 + 3*x2 + 5*x3 + 2*x4 <= 10, "Constraint 1"
prob += x3 + x4 <= 1, "Constraint 2"
prob += x3 - x1 <= 0, "Constraint 3"
prob += x4 - x2 <= 0, "Constraint 4"
prob += x1 == 1, "Constraint S2" # branch from S2
prob += x2 == 1, "Constraint S4" # branch from S4
prob += x4 == 0, "Constraint S7"

# Solve the problem
prob.solve()

# Print the results
print("**** S - 7 : x4 == 0 ****")
print("Status:", pulp.LpStatus[prob.status])
for variable in prob.variables():
    print(f"{variable.name} = {variable.varValue}")
print("Objective value:", pulp.value(prob.objective))

**** S - 7 : x4 == 0 ****
Status: Optimal
x1 = 1.0
x2 = 1.0
x3 = 0.2
x4 = 0.0
Objective value: 15.2


S - 7 `x4 = 0`:
- Fractional
- Branch on x3
- Add x3 = 0 (S9)
- Add x3 = 1 (S10)

#### S8 - Add `x4 = 1`

In [ ]:
# Code for Solving the Problem with Constraint S8
import pulp

# Create a LP maximization problem
prob = pulp.LpProblem("Maximization_Problem", pulp.LpMaximize)

# Define decision variables
x1 = pulp.LpVariable('x1', lowBound=0, upBound=1, cat='Continuous')
x2 = pulp.LpVariable('x2', lowBound=0, upBound=1, cat='Continuous')
x3 = pulp.LpVariable('x3', lowBound=0, upBound=1, cat='Continuous')
x4 = pulp.LpVariable('x4', lowBound=0, upBound=1, cat='Continuous')

# Objective function
prob += 9*x1 + 5*x2 + 6*x3 + 4*x4, "Objective"

# Constraints
prob += 6*x1 + 3*x2 + 5*x3 + 2*x4 <= 10, "Constraint 1"
prob += x3 + x4 <= 1, "Constraint 2"
prob += x3 - x1 <= 0, "Constraint 3"
prob += x4 - x2 <= 0, "Constraint 4"
prob += x1 == 1, "Constraint S2" # branch from S2
prob += x2 == 1, "Constraint S4" # branch from S4
prob += x4 == 1, "Constraint S8"

# Solve the problem
prob.solve()

# Print the results
print("**** S - 8 : x4 == 1 ****")
print("Status:", pulp.LpStatus[prob.status])
for variable in prob.variables():
    print(f"{variable.name} = {variable.varValue}")
print("Objective value:", pulp.value(prob.objective))

**** S - 8 : x4 == 1 ****
Status: Infeasible
x1 = 1.0
x2 = 0.66666667
x3 = 0.0
x4 = 1.0
Objective value: 16.33333335


S - 8 `x4 = 1`:
- Infeasible
- Prune this branch

#### S9 - Add `x3 = 0`

In [ ]:
# Code for Solving the Problem with Constraint S9
import pulp

# Create a LP maximization problem
prob = pulp.LpProblem("Maximization_Problem", pulp.LpMaximize)

# Define decision variables
x1 = pulp.LpVariable('x1', lowBound=0, upBound=1, cat='Continuous')
x2 = pulp.LpVariable('x2', lowBound=0, upBound=1, cat='Continuous')
x3 = pulp.LpVariable('x3', lowBound=0, upBound=1, cat='Continuous')
x4 = pulp.LpVariable('x4', lowBound=0, upBound=1, cat='Continuous')

# Objective function
prob += 9*x1 + 5*x2 + 6*x3 + 4*x4, "Objective"

# Constraints
prob += 6*x1 + 3*x2 + 5*x3 + 2*x4 <= 10, "Constraint 1"
prob += x3 + x4 <= 1, "Constraint 2"
prob += x3 - x1 <= 0, "Constraint 3"
prob += x4 - x2 <= 0, "Constraint 4"
prob += x1 == 1, "Constraint S2" # branch from S2
prob += x2 == 1, "Constraint S4" # branch from S4
prob += x4 == 0, "Constraint S7" # branch from S7
prob += x3 == 0, "Constraint S9" 

# Solve the problem
prob.solve()

# Print the results
print("**** S - 9 : x3 == 0 ****")
print("Status:", pulp.LpStatus[prob.status])
for variable in prob.variables():
    print(f"{variable.name} = {variable.varValue}")
print("Objective value:", pulp.value(prob.objective))

**** S - 9 : x3 == 0 ****
Status: Optimal
x1 = 1.0
x2 = 1.0
x3 = 0.0
x4 = 0.0
Objective value: 14.0


S - 9 `x3 = 0`:
- Integer
- No further branching required

#### S10 - Add `x3 = 1`

In [ ]:
# Code for Solving the Problem with Constraint S10
import pulp

# Create a LP maximization problem
prob = pulp.LpProblem("Maximization_Problem", pulp.LpMaximize)

# Define decision variables
x1 = pulp.LpVariable('x1', lowBound=0, upBound=1, cat='Continuous')
x2 = pulp.LpVariable('x2', lowBound=0, upBound=1, cat='Continuous')
x3 = pulp.LpVariable('x3', lowBound=0, upBound=1, cat='Continuous')
x4 = pulp.LpVariable('x4', lowBound=0, upBound=1, cat='Continuous')

# Objective function
prob += 9*x1 + 5*x2 + 6*x3 + 4*x4, "Objective"

# Constraints
prob += 6*x1 + 3*x2 + 5*x3 + 2*x4 <= 10, "Constraint 1"
prob += x3 + x4 <= 1, "Constraint 2"
prob += x3 - x1 <= 0, "Constraint 3"
prob += x4 - x2 <= 0, "Constraint 4"
prob += x1 == 1, "Constraint S2" # branch from S2
prob += x2 == 1, "Constraint S4" # branch from S4
prob += x4 == 0, "Constraint S7" # branch from S7
prob += x3 == 1, "Constraint S10" 

# Solve the problem
prob.solve()

# Print the results
print("**** S - 10 : x3 == 1 ****")
print("Status:", pulp.LpStatus[prob.status])
for variable in prob.variables():
    print(f"{variable.name} = {variable.varValue}")
print("Objective value:", pulp.value(prob.objective))

**** S - 10 : x3 == 1 ****
Status: Infeasible
x1 = 1.0
x2 = -0.33333333
x3 = 1.0
x4 = 0.0
Objective value: 13.33333335


S - 10 `x3 = 1`:
- Infeasible
- Prune this branch